### DISCUSSION TITLE
PWMAudio use

### DISCUSSION THREAD

**Discussion by pierrotm777** [earlephilhower/arduino-pico#2723] [General]
Hello, I use a RP2040 Waveshare RP2040 Zero and i love you PWMAudio library.
I play with and i can use the demo with success.
I try to build a RC motor simulator.
For that , i want to play two differents sounds, one for the start and another as loop.
I should like to know if it's possible to change the volume, the pitch and the speed of the sample.
Actually, we need to compile a file.h but i should like to select a wav file from a SD card.
Is it possible to add these options ?
I send you my actual code.

Thanks for your fantastic work,
[Engine_Audio.zip](https://github.com/user-attachments/files/18270506/Engine_Audio.zip)

Pierre

**Comment by earlephilhower** [comment]
You can read data from wherever you like to send to PWMAudio, just look at the SD examples and work from there.

To change pitch you can, within reason, just change the PWM frequency.  Higher frequency = higher pitch.  --edit-- Actually, just remembered that changing to non-standard frequencies takes a long time because the Pico has to try and calculate the best new set of dividers which it does using brute-force.  Instead, just drop or duplicate every X samples to drop or raise pitch. 

To change volume you need to scale the samples up or down (i.e. multiply by >1 or <1).  I would recommend using fixed point math instead of `float` since the 2040 has no FPU.

**Comment by pierrotm777** [comment]
[NotSoSimpleMP3Shuffle.zip](https://github.com/user-attachments/files/18270679/NotSoSimpleMP3Shuffle.zip)
I have found this example witch use your BackgroundAudio lib.
Can you give me example code for change volume and PWM frequency in my code please ?

**Reply by earlephilhower** [reply]
Sorry, you're going to have to do your frequency change on your own.  For volume, [BackgroundAudio](https://github.com/earlephilhower/BackgroundAudio) has `setGain` which does what it says on the tin...

**Reply by pierrotm777** [reply]
Ok, but can you give me just the code for change the frenquency ?
By example, if my sample is a 16000, can i write:
audio.begin(20000);
For the gain, i have found, but what is the maximum and the minimum gain ?

**Comment by pierrotm777** [comment]
I don't understand why i haven't sound on my pin 1 :-(
I have no issue with your [PlayRaw](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/PWMAudio/examples/PlayRaw/PlayRaw.ino) example.

**Comment by pierrotm777** [comment]
Hello,
I have found how to play MP3 and WAV,
I have found how to change gain with setGain(x);
I use 0.1 to 1.5 with success.
About the pitch setPWMFrequency isn't very the good solution.
I have tried 10000 to 80000 but i should like bigger bass .
Now i look for how to change the speed.
Can you give me an idea please ?
I send you my example, i have an issue with wav witch start and stop immediatly, can you say me where is my error please ?
[NotSoSimpleWAVShuffle.zip](https://github.com/user-attachments/files/18274964/NotSoSimpleWAVShuffle.zip)
[NotSoSimpleMP3Shuffle.zip](https://github.com/user-attachments/files/18274965/NotSoSimpleMP3Shuffle.zip)

**Reply by Andy2No** [reply]
For example, let's say you have a sample buffer which is 1024 samples of a single cycle of a sine wave.  If you want to play it at 1kHz, you need to read the whole thing out one thousand times per second, but that's 1024000 samples per second.  If your audio playback frequency is only 44.1kHz you need to skip a lot of the samples as you go through the buffer.  Each time you read from the buffer you need to advance the position by 1024000 / 44100, which is about 23.22.

If you only advance by 23 positions each time, you will not get through the entire cycle in 1ms, so the sound will play flat.  You can either try to use floating point or fixed point - e.g. use a 32 bit number where the top 16 bits are the whole number, and the lower 16 bits are the detailed part.  When you're reading a sample you use the whole number part, but you keep the detailed part for the calculations.

The RP2040 is better at fixed point than floating point.  Fixed point is usually faster, even on a CPU with its own FPU, like the RP2350, but it might still be acceptable to use floating point on either, depending on how much else your program needs to get done on each playback cycle.

So, you have a loop which advances the sample position in the buffer, then starts again when it reaches the end. You can pre-calculate the amount it advances by (let's call that delta).  In the loop you only need to do a floating point or fixed point addition:

position= position + delta

You then use the integer part of position as the buffer or array index for your sample.  In fixed point that looks like this:

index = position >> precision; // e.g. precision might be 16 bits

To calculate delta in floating point it's just a division, if necessary casting things as float - you won't need double precision and that's slower.  In fixed point you just shift the numerator (number to be divided by) left by the correct number of bits, before you do the division.  E.g. to represent 2/3 in fixed point you do this:

fixed= (2 << precision) / 3;

So, the calculation for delta is something like the number of samples in the buffer multiplied by the frequency of the note you want, divided by the number of samples per second of playback.  You always do the division part last, to keep the most precision.  The loop just adds delta to the last position, checks for having reached the end and calculates the array or buffer index from the position if not (dropping everything after the decimal point).

Looping shouldn't be by setting the position to zero - you should take off an amount that corresponds to one complete cycle (i.e. subtract the number of samples in the buffer from the position, keeping the part after the decimal point), which might not take you all the way back to zero, but takes you back to near the beginning of the buffer.  Looping back to zero would add a glitch on each cycle.

**Reply by Andy2No** [reply]
If all you can do is put a complete song into a buffer and have it played by the supplied PWMAudio and BackgroundAudioWAV objects, you may have to copy just some of the samples into the playback buffer, using a similar process.  Stretching would work the same way - the position in the source buffer would advance by less than one whole cycle of the playback buffer, as you copied it, some samples would be copied into more than one consecutive playback buffer position.

I haven't looked at those classes to see if they have more capabilities, but that seems like a way of doing it - you go through each position in the playback buffer and advance the position in the source buffer by a calculated delta, which is not required to be a whole number.  Just do that at the start of playing each song.  The source buffer can just be an array - no need for a custom made object.

Getting more bass sounds like a hardware problem, if the recordings don't lack bass.  You can get audio out of a Pi Pico with just a resistor to each channel of a headphones socket.  That allows DC to pass, which gives the maximum possible bass without boosting that part of the signal.  10k to 100k is fine, if you're using a preamp that has fairly high input impedance - I use a FiiO E05 or E06 headphones amplifier.

Technically, that makes a signal that's a little bit too "hot", because it has about a 3V swing, and you ought to limit it to about 2V, for line level, though the FiiO amps don't care either way.  You can do that with a resistive divider, e.g. 10k from the pin to the socket then 20k from the socket to ground.

This produces a signal with a DC offset but that's okay because all preamps and amps have capacitive coupling on the input - which also blocks low bass, so if you still don't get enough, it's caused by the preamp, amp or other audio equipment you're using.

**Comment by pierrotm777** [comment]
Thanks for your answer, but i am not enough good coder for write what you say me.
Can you give me an example of code for change the speed of the sound or the pitch please ?
